### ISSUE TITLE
matplotlib problem

### ISSUE THREAD

**Issue by poettler-ric** [py2exe/py2exe#71]
I am trying to package matplotlib with py2exe.

The `matplotlib.get_py2exe_datafiles()` function (from the wiki) in matplotlib doesn't exist any more, so I used the following code to assemble my data files:

```
import matplotlib as mpl
mpl_files = []
for e in os.walk(mpl.get_data_path()):
    for f in e[2]:
        mpl_files.append(os.path.join(e[0], f))
```

And add it with

```
...
setup(
    name="trainer",
    console=["start.py"],
    data_files=[
        ("mpl-data", mpl_files),
...
```

Which seems to create the necessary files.
But when I then execute the binary I get the following error:

```
<frozen zipimport>:259: MatplotlibDeprecationWarning: Matplotlib installs where the data is not in the mpl-data subdirectory of the package are deprecated since 3.2 and support for them will be removed two minor releases later.
Traceback (most recent call last):
  File "start.py", line 5, in <module>
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 627, in _load_backward_compatible
  File "<frozen zipimport>", line 259, in load_module
  File "hydraulicstrainer\__init__.pyc", line 14, in <module>
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 627, in _load_backward_compatible
  File "<frozen zipimport>", line 259, in load_module
  File "hydraulicstrainer\demo.pyc", line 7, in <module>
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 627, in _load_backward_compatible
  File "<frozen zipimport>", line 259, in load_module
  File "matplotlib\__init__.pyc", line 914, in <module>
  File "matplotlib\cbook\__init__.pyc", line 480, in _get_data_path
  File "matplotlib\__init__.pyc", line 239, in wrapper
  File "matplotlib\__init__.pyc", line 534, in get_data_path
  File "matplotlib\__init__.pyc", line 239, in wrapper
  File "matplotlib\__init__.pyc", line 563, in _get_data_path
NameError: name 'defaultParams' is not defined
```

Did I miss something?

I am using py2exe version 0.10.2.1.

**Comment by albertosottile** [comment]
Hello,
looks like `matplotlib` changed a lot in the way they pack and look at data files so, I fear this will require some time. I will start working on this as soon as I can.

In the meantime, could you provide me a minimal example code to test if `matplotlib` is correctly packed? This includes both a running script and a setup script, and ideally should work also on CI (not so easy). Thank you.

**Comment by poettler-ric** [comment]
Here ist tye python script `mpl_test.py`:
```
import matplotlib as mpl

mpl.use("Agg")

import matplotlib.pyplot as plt

if __name__ == "__main__":
    fig, ax = plt.subplots()
    ax.plot(range(10))
    fig.savefig("test.png", format="png")
```
`setup.py`:
```
import os
import os.path
from distutils.core import setup

import matplotlib as mpl
import py2exe

mpl_files = []
for e in os.walk(mpl.get_data_path()):
    for f in e[2]:
        mpl_files.append(os.path.join(e[0], f))

setup(
    name="mpl_test",
    console=["mpl_test.py"],
    data_files=[
        ("mpl-data", mpl_files),
    ],
)
```
After building it with `python setup.py py2exe` and executing the binary I get the following error message:
```
<frozen zipimport>:259: MatplotlibDeprecationWarning: Matplotlib installs where the data is not in the mpl-data subdirectory of the package are deprecated since 3.2 and support for them will be removed two minor releases later.
Traceback (most recent call last):
  File "mpl_test.py", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 627, in _load_backward_compatible
  File "<frozen zipimport>", line 259, in load_module
  File "matplotlib\__init__.pyc", line 914, in <module>
  File "matplotlib\cbook\__init__.pyc", line 480, in _get_data_path
  File "matplotlib\__init__.pyc", line 239, in wrapper
  File "matplotlib\__init__.pyc", line 534, in get_data_path
  File "matplotlib\__init__.pyc", line 239, in wrapper
  File "matplotlib\__init__.pyc", line 563, in _get_data_path
NameError: name 'defaultParams' is not defined%
```

**Comment by albertosottile** [comment]
**TL;DR:** this is actually a bug. At the end of the comment there is a CI link to download prerelease wheels that should fix it. Please gave those a try and report here if you succeed or fail. Thank you!

First of all, thank you for having reported this. Second of all, I can confirm you this is a bug on the `matplotlib` and `py2exe` sides and that you are not doing anything wrong. I spent the last two evenings on this issue and now I should have a pretty solid understanding of what is happening here and, hopefully, a solution to your problems.

From version 3.2, `matplotlib` started to deprecate installations "where the data is not in the `mpl-data` subdirectory of the package". Of course, this includes also `py2exe`, as we put the package in the zip file and the data files in a folder besides the executable. (IMHO, this was not a really nice move from `matplotlib`, as it cripples all the tools to freeze Python code).

A deprecation warning is never nice but, it should not have prevented `py2exe` from working entirely. In reality, though, they made a mistake in writing the fallback for `if getattr(sys, 'frozen', None)`, which has a case that covers also for `py2exe`: they try to update a dictionary before it is actually defined (the `defaultParams` shown in your last traceback, see https://github.com/matplotlib/matplotlib/issues/18180). A possible solution to this bug could be to open an issue in the `matplotlib` tracker and ask them to fix their fallback code but, a) there's no guarantee they will do that, and b) it will still hold only for the next minor release, and will disappear after, bringing us back again to square one.

Therefore, I tried to address this solely on `py2exe` side. Normally, this kind of issues are dealt in `hooks` by adding a bootcode monkey patch. This bootcode is executed when the packed application is started to patch some modules before actually running them, right after their import, allowing us to use them from the packed executable itself. However, this approach is not viable here, as the rogue code is in `matplotlib.__init__`, effectively making impossible to import it at all from the packed executable. 

So, I had to resort in patching the code at `py2exe` runtime (a.k.a. when creating the packed executable). I still used the hook but not to create a new bootcode, but to patch the `matplotlib` code itself. I will spare you the technicalities but basically, I made use of `ast` to find the `__init__._get_data_path` function and replace it with something suitable for `py2exe`, then compile the patched tree and embed this patched bytecode in the zip file.

The results of this effort made the test you posted here pass in CI. Actually, I used a `setup.py` script that does not try to embed the full package, as that could also cause errors. Also, you should not copy the data directory manually, as `py2exe`takes care of this for you. Here is the script that I encourage you to use:
```python
from distutils.core import setup
import py2exe

setup(console=[{ "script": "matplotlib_test.py"}])
```

The same CI job produces updated test wheels that include this fix that you can find at the bottom of the run page here: https://github.com/py2exe/py2exe/actions/runs/625886951

It would be great if you could download the relevant wheel for you, install it on your system, try again your test with the updated setup script reported above, and confirm if this fixed the error also for you. In case of positive feedback, I will include this fix in the next release of `py2exe`.

I hope all this helped. Thanks for using `py2exe`.

Side note: as you mentioned, documentation on the website is extremely outdated, and should be definitely updated. However, I barely have the time to somewhat maintain the code, and I cannot possibly update also that documentation. If you are willing to edit the website wiki with what you learned from this issue, I would really appreciate that. Thank you in advance.

**Comment by poettler-ric** [comment]
Thanks for your effort. It works for the trimmed down problem and also with my original project.
